eggtimer.js

```diff
@@ -1,326 +1,681 @@
+
+    let labelsParams = commonParams();
+    labelsParams.number = prContext.pr.number;
+    let labelsPromise = getLabels(labelsParams);
+
+    Promise.all([reviewsPromise, statusPromise, mergeablePromise, labelsPromise]).then((results) => {
+            if (prContext.signaled)
+                return Promise.reject("signaled");
+
+            let reviews = results[0];
+            let statuses = results[1];
+            let mergeable = results[2].mergeable;
+            let labels = results[3];
+
+            if ((mergeable !== true) || !approved(reviews) || !allChecksSuccessful(statuses))
+                return Promise.reject("not ready for merging yet");
+
+            if (markedAsMerged(labels))
+                return Promise.reject("already merged");
+
+            if (markedAsFailed(labels))
+                console.log("PR" + prContext.pr.number + " previous merge was unsuccessful");
+
+            checkBeforeStartMerging(prContext);
+            return Promise.resolve(true);
+    })
+    .catch((err) => {
+        console.error("Can't merge PR" + prContext.pr.number + ":", err);
+        processNextPR();
+    });
+}
+
+function checkBeforeStartMerging(prContext) {
+   let refParams = commonParams();
+   refParams.ref = mergingTag(prContext.pr.number);
+   let tagTreeSha = null;
+   prContext.tagSha = null;
+   console.log("Checking whether to skip due to a previous unsuccessful merge for PR" + prContext.pr.number);
+   getReference(refParams)
+       .then( (sha) => {
+           let params = commonParams();
+           params.sha = sha;
+           prContext.tagSha = sha;
+           let commitPromise = getCommit(params);
+
+           let prParams = commonParams();
+           prParams.number = prContext.pr.number;
+           let prPromise = getPR(prParams);
+
+           return Promise.all([commitPromise, prPromise]);
+       })
+       .then((results) => {
+           let commitObj = results[0];
+           let pr = results[1];
+           assert(pr.number = prContext.pr.number);
+           let params = commonParams();
+           tagTreeSha = commitObj.treeSha;
+           params.ref = "pull/" + prContext.pr.number + "/merge";
+           return getReference(params);
+       })
+       .then((sha) => {
+           let params = commonParams();
+           params.sha = sha;
+           return getCommit(params);
+       })
+       .then((obj) => {
+           if (obj.treeSha !== tagTreeSha)
+               return Promise.reject("merge sha has changed");
+           let statusParams = commonParams();
+           statusParams.ref = prContext.tagSha;
+           return getStatuses(statusParams);
+       })
+       .then((checks) => {
+           if (!checkValues(checks, Config.checks_number)) {
+               processNextPR();
+               return Promise.resolve("Some auto_branch checks failed for PR" + prContext.pr.number);
+           }
+           return Promise.reject("all checks succeeded");
```

Comment:
The overall logic of my 'checkBeforeStartMerging()' makes the following sense to me:  try to check whether we should avoid merging for some reason. If there is a reason, or we can't calculate that reason(an exception emerged), we start merging as usual. The natural way to make this working is to use catch block for both these situations.

Reply:
Please change the overall logic to a sequence of steps towards merging, with exceptions or rejections used to handle merge precondition violations and other problems. AFAICT, once you do that, then "the natural way to make this working" will itself be natural.

In other words, if the correct solution is unnatural, then the wrong problem has been solved.

Reply:
>  try to check whether we should avoid merging for some reason. If [...] we can't calculate that reason(an exception emerged), we start merging as usual.

This part of the original logic seems wrong to me.

* In negative logic: If we cannot be sure that merging is not prohibited, then we should _not_ merge.
* In positive logic: If we cannot be sure that merging is allowed, then we should _not_ merge.

No matter which way you prefer to look at the problem, an exception means "no merging" rather than "merging as usual".